crates/bevy_input_focus/src/tab_navigation.rs

```diff
@@ -0,0 +1,336 @@
+    ///
+    /// If no focusable entities are found, then this function will return either the first
+    /// or last focusable entity, depending on the direction of navigation. For example, if
+    /// `action` is `Next` and no focusable entities are found, then this function will return
+    /// the first focusable entity.
+    pub fn navigate(&self, focus: Option<Entity>, action: NavAction) -> Option<Entity> {
+        // If there are no tab groups, then there are no focusable entities.
+        if self.tabgroup_query.is_empty() {
+            warn!("No tab groups found");
+            return None;
+        }
+
+        // Start by identifying which tab group we are in. Mainly what we want to know is if
+        // we're in a modal group.
+        let tabgroup = focus.and_then(|focus_ent| {
+            self.parent_query
+                .iter_ancestors(focus_ent)
+                .find_map(|entity| {
+                    self.tabgroup_query
+                        .get(entity)
+                        .ok()
+                        .map(|(_, tg, _)| (entity, tg))
+                })
+        });
+
+        if focus.is_some() && tabgroup.is_none() {
+            warn!("No tab group found for focus entity");
+            return None;
+        }
+
+        self.navigate_in_group(tabgroup, focus, action)
+    }
+
+    fn navigate_in_group(
+        &self,
+        tabgroup: Option<(Entity, &TabGroup)>,
+        focus: Option<Entity>,
+        action: NavAction,
+    ) -> Option<Entity> {
+        // List of all focusable entities found.
+        let mut focusable: Vec<(Entity, TabIndex)> =
+            Vec::with_capacity(self.tabindex_query.iter().len());
+
+        match tabgroup {
+            Some((tg_entity, tg)) if tg.modal => {
+                // We're in a modal tab group, then gather all tab indices in that group.
+                if let Ok((_, _, children)) = self.tabgroup_query.get(tg_entity) {
+                    for child in children.iter() {
+                        self.gather_focusable(&mut focusable, *child);
+                    }
+                }
+            }
+            _ => {
+                // Otherwise, gather all tab indices in all non-modal tab groups.
+                let mut tab_groups: Vec<(Entity, TabGroup)> = self
+                    .tabgroup_query
+                    .iter()
+                    .filter(|(_, tg, _)| !tg.modal)
+                    .map(|(e, tg, _)| (e, *tg))
+                    .collect();
+                // Stable sort by group order
+                tab_groups.sort_by(compare_tab_groups);
+
+                // Search group descendants
+                tab_groups.iter().for_each(|(tg_entity, _)| {
+                    self.gather_focusable(&mut focusable, *tg_entity);
+                });
+            }
+        }
+
+        if focusable.is_empty() {
+            warn!("No focusable entities found");
```

Comment:
nit: This should be a debug! or detailed_trace! otherwise it would be difficult to understand without context.

Another option is return an `Result` from this function instead of `Option`.